pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,538 @@
+// Copyright 2021 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package stanpubsub provides a pubsub implementation for NATS Streaming client.
+// Use OpenTopic to construct a *pubsub.Topic, and/or OpenSubscription to construct a
+// *pubsub.Subscription. This package uses gob to encode and decode driver.Message to
+// []byte.
+//
+// URLs
+//
+// For pubsub.OpenTopic and pubsub.OpenSubscription, stanpubsub registers
+// for the scheme "stan".
+// The default URL opener will connect to a default server based on the environment variables:
+// 	- STAN_SERVER_URL	- url address of the NATS server
+//	- STAN_CLUSTER_ID	- NATS Streaming cluster id
+//	- STAN_CLIENT_ID	- NAST Streaming client id
+// To customize the URL opener, or for more details on the URL format,
+// see URLOpener.
+// See https://gocloud.dev/concepts/urls/ for background information.
+//
+// Message Delivery Semantics
+//
+// NATS Streaming supports at-least-once semantics; by default applications doesn't need to call Message.Ack,
+// and must not call Message.Nack.
+// Optionally a user can set a subscription to be manually acked by using `stan.SetManualAckMode` option
+// or by setting `manualAck` query parameter when setting from URL.
+// See https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery
+// for more background.
+//
+// As
+//
+// stanpubsub exposes the following types for As:
+//  - Topic: stan.Conn
+//  - Subscription: stan.Subscription
+//  - Message.BeforeSend: None.
+//  - Message.AfterSend: None.
+//  - Message: *stan.Msg
+
+package stanpubsub
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/stan.go"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/batcher"
+	"gocloud.dev/pubsub/driver"
+)
+
+func init() {
+	dd = new(defaultDialer)
+	pubsub.DefaultURLMux().RegisterTopic(Scheme, dd)
+	pubsub.DefaultURLMux().RegisterSubscription(Scheme, dd)
+}
+
+// This defaultDialer variable is used only for the test purpose.
```

Comment:
Nit: "for testing purposes."

Reply:
Removed that variable with a suggested `connect` function implementation.